File: enode/src/main/java/org/enodeframework/domain/Repository.java
package org.enodeframework.domain;

import java.util.concurrent.CompletableFuture;

public interface Repository {
    /**
     * Get an aggregate from memory cache, if not exist, get it from event store.
     */
    <T extends AggregateRoot> CompletableFuture<T> getAsync(Class<T> aggregateRootType, Object aggregateRootId);

    /**
     * Get an aggregate from memory cache, if not exist, get it from event store.
     */
    CompletableFuture<AggregateRoot> getAsync(Object aggregateRootId);

    /**
     * Refresh memory cache with given aggregate.
     */
    <T extends AggregateRoot> void refreshAggregate(T aggregateRoot);
}

File: enode/src/main/java/org/enodeframework/domain/Repository.kt
package org.enodeframework.domain

import java.util.concurrent.CompletableFuture

interface Repository {
    /**
     * Get an aggregate from memory cache, if not exist, get it from event store.
     */
    fun <T : AggregateRoot?> getAsync(aggregateRootType: Class<T>, aggregateRootId: String): CompletableFuture<T>

    /**
     * Refresh memory cache with given aggregate.
     */
    fun <T : AggregateRoot?> refreshAggregate(aggregateRoot: T)
}
